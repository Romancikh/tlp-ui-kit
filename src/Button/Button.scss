.button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  outline: 0;
  border: 0;
  margin: 0;
  cursor: pointer;
  user-select: none;
  vertical-align: middle;
  text-decoration: none;
  appearance: none;
  font-family: Roboto, Helvetica, Arial, sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 16px;
  border-radius: 4px;
  transition:
    background-color 250ms cubic-bezier(0.4, 0, 0.2, 1),
    box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1),
    border-color 250ms cubic-bezier(0.4, 0, 0.2, 1),
    color 250ms cubic-bezier(0.4, 0, 0.2, 1);
  overflow: hidden;

  span.ripple {
    position: absolute;
    border-radius: 50%;
    transform: scale(0);
    animation: ripple 0.6s ease-in;
  }

  &:disabled {
    pointer-events: none;
    cursor: default;
  }

  &_variant_contained {
    color: #fff;
    box-shadow:
      rgba(0, 0, 0, 0.2) 0px 3px 1px -2px,
      rgba(0, 0, 0, 0.14) 0px 2px 2px 0px,
      rgba(0, 0, 0, 0.12) 0px 1px 5px 0px;
    background-color: #1976d2;

    span.ripple {
      background: #ffffff50;
    }

    &:disabled {
      box-shadow: none;
      color: rgba(0, 0, 0, 0.26);
      background-color: rgba(0, 0, 0, 0.12);
    }

    &:hover {
      background-color: #1565c0;
      box-shadow:
        0px 2px 4px -1px rgba(0, 0, 0, 0.2),
        0px 4px 5px 0px rgba(0, 0, 0, 0.14),
        0px 1px 10px 0px rgba(0, 0, 0, 0.12);
    }

    &.button_color_success {
      background-color: #2e7d32;

      &:hover {
        background-color: #1b5e20;
      }
    }
  }

  &_variant_outlined {
    background-color: transparent;
    padding: 5px 15px;
    border: 1px solid #1976d280;
    color: #1976d2;

    span.ripple {
      background: #1976d250;
    }

    &:hover {
      background-color: #1976d20a;
      border: 1px solid #1976d2;
    }

    &:disabled {
      color: #00000042;
      border: 1px solid #0000001f;
      &:hover {
        background: transparent;
      }
    }

    &.button_color_success {
      border: 1px solid #2e7d3280;
      color: #2e7d32;

      span.ripple {
        background: #2e7d3250;
      }

      &:hover {
        background-color: #2e7d320a;
        border: 1px solid #2e7d32;
      }
    }
  }

  &__start-icon {
    display: inherit;
    margin-right: 8px;
    margin-left: -4px;
    font-size: 20px !important;
  }
}

@keyframes ripple {
  to {
    transform: scale(4);
    opacity: 0;
  }
}
