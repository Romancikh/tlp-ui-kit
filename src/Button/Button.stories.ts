import type { Meta, StoryObj } from "@storybook/react";

import Button from "./Button";

const meta = {
  title: "Button",
  component: Button,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  args: {
    variant: "contained",
    disabled: false,
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Contained: Story = {
  args: {
    variant: "contained",
    children: "contained",
  },
};

export const Outlined: Story = {
  args: {
    variant: "outlined",
    children: "outlined",
  },
};

export const Disabled: Story = {
  args: {
    variant: "contained",
    disabled: true,
    children: "disabled",
  },
};

export const Success: Story = {
  args: {
    variant: "contained",
    color: "success",
    children: "success",
  },
};
