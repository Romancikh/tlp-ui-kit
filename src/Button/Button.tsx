import { CSSProperties, MouseEvent, PropsWithChildren, ReactNode } from "react";

import "./Button.scss";

type ButtonProps = PropsWithChildren & {
  alignSelf?: "flex-end";
  color?: "success";
  disabled?: boolean;
  marginTop?: number;
  minWidth?: "unset";
  onClick?: () => void;
  padding?: number;
  startIcon?: ReactNode;
  variant: "contained" | "outlined";
};

function Button({
  alignSelf,
  children,
  color,
  disabled,
  marginTop,
  minWidth,
  onClick,
  padding,
  startIcon,
  variant,
}: ButtonProps) {
  const handleRipple = (event: MouseEvent<HTMLButtonElement>) => {
    const button = event.currentTarget;
    const rippleElement = document.createElement("span");
    const rect = button.getBoundingClientRect();
    const size = Math.max(rect.width, rect.height);
    const x = event.clientX - rect.left - size / 2;
    const y = event.clientY - rect.top - size / 2;

    rippleElement.style.width = rippleElement.style.height = `${size}px`;
    rippleElement.style.left = `${x}px`;
    rippleElement.style.top = `${y}px`;

    rippleElement.classList.add("ripple");

    button.appendChild(rippleElement);

    setTimeout(() => {
      rippleElement.remove();
    }, 600);
  };

  const buttonStyle = {
    alignSelf,
    marginTop,
    minWidth,
    padding,
  } satisfies CSSProperties;

  return (
    <button
      className={`button button_variant_${variant} button_color_${color}`}
      disabled={disabled}
      onClick={onClick}
      onMouseDown={handleRipple}
      style={buttonStyle}
      tabIndex={0}
      type="button"
    >
      {startIcon && <span className="button__start-icon">{startIcon}</span>}
      {children}
    </button>
  );
}

export default Button;
