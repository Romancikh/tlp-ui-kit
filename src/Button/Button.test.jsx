import { render } from "@testing-library/react";
import Button from "./Button";

test("Contained button renders correctly", () => {
  const { container } = render(
    <Button variant="contained" children="contained" />,
  );
  expect(container).toMatchSnapshot();
});

test("Outlined button renders correctly", () => {
  const { container } = render(
    <Button variant="outlined" children="outlined" />,
  );
  expect(container).toMatchSnapshot();
});

test("Disabled button renders correctly", () => {
  const { container } = render(
    <Button variant="Disabled" children="Disabled" />,
  );
  expect(container).toMatchSnapshot();
});

test("Success button renders correctly", () => {
  const { container } = render(
    <Button variant="contained" color="success" children="Success" />,
  );
  expect(container).toMatchSnapshot();
});
