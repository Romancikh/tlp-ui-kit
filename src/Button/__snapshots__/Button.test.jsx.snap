// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contained button renders correctly 1`] = `
<div>
  <button
    class="button button_variant_contained button_color_undefined"
    tabindex="0"
    type="button"
  >
    contained
  </button>
</div>
`;

exports[`Disabled button renders correctly 1`] = `
<div>
  <button
    class="button button_variant_Disabled button_color_undefined"
    tabindex="0"
    type="button"
  >
    Disabled
  </button>
</div>
`;

exports[`Outlined button renders correctly 1`] = `
<div>
  <button
    class="button button_variant_outlined button_color_undefined"
    tabindex="0"
    type="button"
  >
    outlined
  </button>
</div>
`;

exports[`Success button renders correctly 1`] = `
<div>
  <button
    class="button button_variant_contained button_color_success"
    tabindex="0"
    type="button"
  >
    Success
  </button>
</div>
`;
